modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C

```diff
@@ -0,0 +1,286 @@
+        if (mesh_node_id_to_boundary_node_id.find(old_node_id) ==
+            mesh_node_id_to_boundary_node_id.end())
+          mooseError("Node id", old_node_id, " not linked to new node id.");
+        dof_id_type new_node_id = mesh_node_id_to_boundary_node_id.find(old_node_id)->second;
+        new_bnd_elem->set_node(j) = boundary_nodes[new_node_id];
+      }
+    }
+  }
+
+  // partition the boundary mesh
+  boundary_mesh->prepare_for_use();
+  setPartitionerHelper(boundary_mesh);
+  boundary_mesh->partition(_n_patches);
+
+  // prepare sideset names and boundary_ids added to mesh
+  std::vector<BoundaryName> sideset_names =
+      sidesetNameHelper(boundary_info.get_sideset_name(_sideset));
+
+  std::vector<boundary_id_type> boundary_ids =
+      MooseMeshUtils::getBoundaryIDs(*mesh, sideset_names, true);
+
+  mooseAssert(sideset_names.size() == _n_patches,
+              "sideset_names must have as many entries as user-requested number of patches.");
+  mooseAssert(boundary_ids.size() == _n_patches,
+              "boundary_ids must have as many entries as user-requested number of patches.");
+
+  // loop through all elements in the boundary mesh and assign the side of
+  // the _original_ element to the new sideset
+  for (const auto & elem : boundary_mesh->active_element_ptr_range())
+  {
+    if (boundary_elem_to_mesh_elem.find(elem->id()) == boundary_elem_to_mesh_elem.end())
+      mooseError("Element in the boundary mesh with id ",
+                 elem->id(),
+                 " not found in boundary_elem_to_mesh_elem.");
+
+    auto side = boundary_elem_to_mesh_elem.find(elem->id())->second;
+
+    mooseAssert(elem->processor_id() < boundary_ids.size(),
+                "Processor id larger than number of patches.");
+    boundary_info.add_side(
+        std::get<0>(side), std::get<1>(side), boundary_ids[elem->processor_id()]);
+  }
+
+  // make sure new boundary names are set
+  for (unsigned int j = 0; j < boundary_ids.size(); ++j)
+  {
+    boundary_info.sideset_name(boundary_ids[j]) = sideset_names[j];
+    boundary_info.nodeset_name(boundary_ids[j]) = sideset_names[j];
+  }
+
+  return mesh;
+}
+
+std::vector<BoundaryName>
+PatchSidesetGenerator::sidesetNameHelper(const std::string & base_name) const
+{
+  std::vector<BoundaryName> rv;
+  for (unsigned int j = 0; j < _n_patches; ++j)
+  {
+    std::stringstream ss;
+    ss << base_name << "_" << j;
+    rv.push_back(ss.str());
+  }
+  return rv;
+}
+
+Elem *
+PatchSidesetGenerator::boundaryElementHelper(std::unique_ptr<libMesh::ReplicatedMesh> & mesh,
+                                             int type) const
+{
+  switch (type)
+  {
+    case 5:
+      return mesh->add_elem(new libMesh::Quad4);
+    default:
+      mooseError("Unsupported element type (libMesh elem_type enum): ", type);
+  }
+}
+
+void
+PatchSidesetGenerator::setPartitionerHelper(std::unique_ptr<libMesh::ReplicatedMesh> & mesh) const
+{
+  switch (_partitioner_name)
```

Comment:
This code is copied from MooseMesh. Why don't we factor that code out into a routine (like this) and call it from both places?

Reply:
It'd have to be a static function because MeshGenerators don't have a MooseMesh.
But the function would have state from this:

```
        if (!isParamValid("centroid_partitioner_direction"))
          mooseError("If using the centroid partitioner you _must_ specify "
                     "centroid_partitioner_direction!");

        MooseEnum direction = getParam<MooseEnum>("centroid_partitioner_direction");
```

Reply:
MeshGenerators do have a reference to the mesh object and that object does exist before the generators run.

Reply:
you're right.